rbd: deprecate .rbd.csi.ceph.com/thick-provisioned metadata key

Madhu-1 opened the pull request:
As image metadata key starting with '.rbd' will not be copied when we do clone or mirroring, deprecating the old key for the same reason use 'csi.ceph.com/thick-provisioned' to set image metadata. set thickProvisionMetaKey only if the parent volume is having a deprecated metadata key set.

updates #2219 
closes:  #2121

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

## Test scenarios

### Backward compatibility

* Create Thick PVC from cephcsi 3.3.0
* Upgrade cephcsi from  3.3.0 to latest(custom build from this PR)
* Create Snapshot from Thick PVC-> Verify metadata with the new key is  present on rbd image created during CreateSnapshot operation)
* Restore Snapshot to a thick PVC -> Verify the metadata with new key
* Create PVC-PVC clone from Thick PVC created during 3.3.0 -> Verify  the metadata with a new key on the clone image

### Mirroring

Able to mount/resize the rbd image after the failover to another cluster.

Madhu-1 committed:
rbd: deprecate .rbd.csi.ceph.com/thick-provisioned metadata key

As image metadata key starting with '.rbd' will not
be copied when we do clone or mirroring, deprecating
the old key for the same reason use
'csi.ceph.com/thick-provisioned' to set image metadata.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
e2e: update e2e to validate thick metadata

updated e2e tests to validate the thick
metadata set on the rbd image for thick PVC.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -1533,10 +1537,14 @@ func (rv *rbdVolume) setThickProvisioned() error {
-	if err != nil {
+	if err == librbd.ErrNotFound {
+		// check if the image is having deprecated metadata key.
+		value, err = ri.GetMetadata(deprecatedthickProvisionMetaKey)
```
If we get here, it may be simpler to set the new key in any case. That way, only one check for the metadata is needed when the image is thin-provisioned.

It also helps with converting old images to the new metadata, and the deprecated key can be used only in this function (or ideally a new helper).

Madhu-1 replied:
@nixpanic if the error  is `err == librbd.ErrNotFound` and the metadata key `thickProvisionMetaKey` either it could be thin volume or it might be thick volume with old metadata set. For converting, yes if the value is "true" then we can set the metadata key again which will convert the old images to new metadata. Is that what you are looking for?

```suggestion
func (ri *rbdImage) isThickProvisioned() (bool, error) {
	value, err := ri.GetMetadata(thickProvisionMetaKey)
	if err == librbd.ErrNotFound {
		// check if the image is having deprecated metadata key.
		value, err = ri.GetMetadata(deprecatedthickProvisionMetaKey)
		if err == librbd.ErrNotFound {
			return false, nil
		}
		if value == "true" {
		      err=vi.setThickProvisioned(nil)
		      if err!=nil{
		            return false,err
		      }
		}...
	}
```

nixpanic replied:
When the image is thin-provisioned, setting `thickProvisionMetaKey` to `false` will prevent a 2nd lookup of the metadata on subsequent usage. That makes it also unneeded for `setThickProvisioned()` to check for a parent volume and makes the rest of this PR simpler.

Eventually we can set the metadata (when thin-provisioned) on image creation, if we decide that has a benefit.

Madhu-1 replied:
Yes, that makes sense but the only thing is we will be setting image metadata on all images irrespective it's thick or thin. Am fine either way let me know the suggestions. but still, we need to pass parent volume to `setThickProvisioned` to support backward compatibility?

nixpanic replied:
Setting the value would not hurt, so if it is not intentionally set (thin-provisioned), setting it to `false` is fine.

I do not think it will be needed to check if a parent volume has the metadata set when detecting the old key and setting the new one. I can not think of a reason why that is needed, as we already explicitly set the metadata when cloning or restoring from a snapshot.

nixpanic replied:
Still waiting on an update for this. There should not be need to check the parent  volume when getting/setting the thick-provisioned key.

Madhu-1 replied:
made changes as per suggestion.

Madhu-1 commented:
Busy with other PR. will revisit this tomorrow

Madhu-1 committed:
rbd: set thick metadata if ThickProvision is set

instead of checking the parent is thick provisioned
or not we can decide based on the rbdVol generated
from the request. If the request is to create a Thick
Image. set metadata without checking the parent.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: make setThickProvisioned as method of rbdImage

isThickProvisioned is already method of the rbdImage
to keep similar thick provisioner related functions
common making isThickProvisioned as method of rbdImage.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: set image metadata in isThickProvisioned

setting metadata in isThickProvisioned method
helps us to avoid checking thick metakey and
deprecated metakey for both thick and thin
provisioned images and also this will easily
help us to migrated the deprecated key to new key.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

nixpanic reviewed (approved):
great, much cleaner! :+1:
